### PR TITLE
[SYCL] Stop reinterpret from changing buffer_allocator type to const

### PR DESCRIPTION
**Pull request by AidanBeltonS**
This PR prevents the `buffer_allocator` from being rebound to a const type. Currently when a `buffer` is reinterpreted from type `T` to type `const T` the `buffer_allocator` type is also changed to `const T`. This does not agree with the SYCL spec which states "A buffer of data type const T uses buffer_allocator<T> by default."

This PR also adds a compile time test which validates the returned type.

### CODE REVIEW

File: sycl/test/basic_tests/buffer/buffer_reinterpret.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
+// RUN: %clangxx -fsycl %s -o %t.out
```
Comment (bader) [inline sycl/test/basic_tests/buffer/buffer_reinterpret.cpp:1]:
```suggestion
// RUN: %clangxx -fsycl -fsyntax-only %s
```
We need only to check that compiler front-end do not report errors, right?

File: sycl/test/basic_tests/buffer/buffer_reinterpret.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+#include <sycl.hpp>
```
Comment (bader) [inline sycl/test/basic_tests/buffer/buffer_reinterpret.cpp:3]:
```suggestion
#include <sycl/sycl.hpp>
```
I think we need to file a bug report for this. I would expect compiler to emit a error for `#include <sycl.hpp>`.

Comment (keryell) [reply sycl/test/basic_tests/buffer/buffer_reinterpret.cpp]: > I think we need to file a bug report for this. I would expect compiler to emit a error for `#include <sycl.hpp>`.

I am unsure about what you are asking.
I would not expect the compiler to emit a specific error for `#include <sycl.hpp>` any further than with any `#include <whatever_not_found.hpp>`.
If someone declares at some point for an other purpose `sycl.hpp` in the include path unrelated to a SYCL impleemntation, `#include <sycl.hpp>` should work in this context, without the compiler suggesting to use `#include <sycl/sycl.hpp>` instead, right?

Comment (bader) [reply sycl/test/basic_tests/buffer/buffer_reinterpret.cpp]: I'm not asking to emit an error for `#include <sycl.hpp>` expression, but rather remove an include path, that allows users to include **standard** SYCL header with just `#include <sycl.hpp>`.

**Comment by AidanBeltonS** [comment]
> Hm, to me the spec is not very clear here:

I agree that the spec is not clear. I also agree that your proposed solution is better than changing how the rebinding occurs.
I have proposed a fix to the SYCL Spec to update this interface as I expect this is a mistake. 
https://github.com/KhronosGroup/SYCL-Docs/pull/286

**Comment by AidanBeltonS** [comment]
@romanovvlad the change to the Khronos spec has been merged, clarifying this issue. I have updated the PR to change the interface rather than the rebind operation. It seems like the ESIMD CI is failing, I am not sure if this is related to my changes. Is the current CI failure a known issue?

**Comment by romanovvlad** [comment]
> @romanovvlad the change to the Khronos spec has been merged, clarifying this issue. I have updated the PR to change the interface rather than the rebind operation. It seems like the ESIMD CI is failing, I am not sure if this is related to my changes. Is the current CI failure a known issue?

I guess CI fails are known issues and they should disappear after https://github.com/intel/llvm-test-suite/commit/7f0e3f0876f483790f47e29e5441683e56cb5497.

**Review by romanovvlad — APPROVED** [review]

**Comment by AidanBeltonS** [comment]
@romanovvlad has the CI been fixed yet? If so could we rerun the tests? Many thanks

**Comment by AidanBeltonS** [comment]
@romanovvlad, @steffenlarsen could this be merged at some point, or is this waiting on more reviews?

**Comment by steffenlarsen** [comment]
Looks like the spec changes have been merged too, so I believe it is ready. Merging.

**Pull request merged**